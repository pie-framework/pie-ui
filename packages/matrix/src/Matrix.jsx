import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import ChoiceInput from './ChoiceInput';
import {Collapsible} from '@pie-lib/render-ui';

const PromptWrapper = styled.div`
  vertical-align: middle;
  color: var(--pie-primary-text, var(--pie-text, #000000));
`;

const MatrixWrapper = styled.div`
  font-family: Roboto, Arial, Helvetica, sans-serif;
`;

const MatrixGridWrapper = styled.div`
  display: grid;
  grid-template-columns: ${props => props.gridTemplateColumns};
  grid-column-gap: 30px;
  grid-row-gap: 20px;
  margin-top: 40px;
`;


const MatrixGridItem = styled.div`
  align-self: center;
  justify-self: center;
  white-space: nowrap;
`;

const Matrix = (props) => {
  const {
    disabled,
    prompt,
    onSessionChange,
    columnLabels,
    matrixValues,
    rowLabels,
    session,
    teacherInstructions
  } = props;

  const gridMatrixItems = [];

  for (let rowIndex = 0; rowIndex < rowLabels.length + 1; rowIndex++) {
    for (let columnIndex = 0; columnIndex < columnLabels.length + 1; columnIndex++) {
      let gridMatrixItem;
      if (rowIndex === 0 && columnIndex === 0) {
        gridMatrixItem = null;
      } else if (rowIndex === 0) {
        gridMatrixItem = columnLabels[columnIndex - 1];
      } else if (columnIndex === 0) {
        gridMatrixItem = rowLabels[rowIndex - 1];
      } else {
        const matrixKey = `${rowIndex - 1}-${columnIndex - 1}`;
        const matrixValue = matrixValues[matrixKey];
        gridMatrixItem = (<ChoiceInput
          matrixKey={matrixKey}
          matrixValue={matrixValue || 0}
          disabled={disabled}
          onChange={onSessionChange}
          checked={session.value && session.value.hasOwnProperty(matrixKey)}
        />);
      }
      gridMatrixItems.push(gridMatrixItem);
    }
  }
  const gridTemplateColumns = [...columnLabels, {}].map(() => 'min-content').join(' ');

  return (
    <MatrixWrapper>
      {teacherInstructions && (
        <Collapsible
          labels={{
            hidden: 'Show Teacher Instructions',
            visible: 'Hide Teacher Instructions',
          }}
        >
          <div dangerouslySetInnerHTML={{__html: teacherInstructions}}/>
        </Collapsible>
      )}
      <br/>
      <br/>
      <PromptWrapper dangerouslySetInnerHTML={{__html: prompt}}/>
      <MatrixGridWrapper gridTemplateColumns={gridTemplateColumns}>
        {gridMatrixItems.map((gridMatrixItem, gridMatrixIndex) => {
          return (
            <MatrixGridItem key={`${gridMatrixIndex}`}>
              {gridMatrixItem}
            </MatrixGridItem>
          );
        })}
      </MatrixGridWrapper>
    </MatrixWrapper>
  );
};

Matrix.propTypes = {
  prompt: PropTypes.string,
  teacherInstructions: PropTypes.string,
  session: PropTypes.object,
  matrixValues: PropTypes.object,
  rowLabels: PropTypes.arrayOf(PropTypes.string),
  columnLabels: PropTypes.arrayOf(PropTypes.string),
  disabled: PropTypes.bool.isRequired,
  onSessionChange: PropTypes.func.isRequired
};

Matrix.defaultProps = {
  session: {
    value: {}
  },
};

export default Matrix;
