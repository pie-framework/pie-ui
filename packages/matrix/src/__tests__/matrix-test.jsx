import React from 'react';
import {shallow} from 'enzyme';
import Matrix from '../Matrix';

describe('Matrix', () => {
  const getShallowMatrix = ((onSessionChange, propsToOverride = {}) => {
    const props = {
      prompt: 'magic prompt',
      teacherInstructions: 'instructions',
      session: {},
      matrixValues: {},
      rowLabels: [],
      columnLabels: [],
      disabled: false,
      onSessionChange,
      ...propsToOverride
    };
    return shallow(<Matrix {...props}/>);
  });

  describe('snapshot', () => {
    it('renders', () => {
      const onSessionChange = jest.fn();
      const shallowMatrix = getShallowMatrix(onSessionChange);
      expect(shallowMatrix).toMatchSnapshot();
    });
  });
});
