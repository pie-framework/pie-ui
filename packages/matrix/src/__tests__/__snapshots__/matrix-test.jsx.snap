// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matrix snapshot renders 1`] = `
<styled.div>
  <Component
    labels={
      Object {
        "hidden": "Show Teacher Instructions",
        "visible": "Hide Teacher Instructions",
      }
    }
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "instructions",
        }
      }
    />
  </Component>
  <br />
  <br />
  <styled.div
    dangerouslySetInnerHTML={
      Object {
        "__html": "magic prompt",
      }
    }
  />
  <styled.div
    gridTemplateColumns="min-content"
  >
    <styled.div
      key="0"
    />
  </styled.div>
</styled.div>
`;
